Imports globals from @types/react

**Bug report**
When using the external-imports flag with react, the tools doesn't seem to recognize that certain types are global and imports them from react in the declaration file. This causes an error during the generated file check.

```
$ dts-bundle-generator -o index.d.ts index.tsx --external-imports react
Compiling input files...
Processing index.tsx
Adding import with name "HTMLButtonElement" for library "react"
Adding import with name "FunctionComponent" for library "react"
Adding import with name "HTMLAttributes" for library "react"
Writing index.tsx -> index.d.ts
Checking generated files...
index.d.ts(1,45): error TS2305: Module '"./node_modules/@types/react"' has no exported member 'HTMLButtonElement'.
Error: Compiled with errors
```

**Input code**

```tsx
import React, { FunctionComponent, HTMLAttributes } from "react";

interface ButtonProps extends HTMLAttributes<HTMLButtonElement> { }

export const Button: FunctionComponent<ButtonProps> = (props) => <button {...props} />;
```

**Expected output**

```ts
import { FunctionComponent, HTMLAttributes } from "react";
export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
}
export declare const Button: FunctionComponent<ButtonProps>;
```

**Actual output**

```ts
import { FunctionComponent, HTMLAttributes, HTMLButtonElement } from 'react';
export interface ButtonProps extends HTMLAttributes<HTMLButtonElement> {
}
export declare const Button: FunctionComponent<ButtonProps>;
```

Hi,

sorry for long answer - I was a little bit busy last week 🙁

Your code doesn't compile in clean env. Here what I did:

1. `mkdir temp`
1. Add `index.tsx` file with your content
1. `npm install typescript @types/react dts-bundle-generator`
1. `./node_modules/.bin/tsc --init`
1. Add `"lib": ["dom", "es5"]` to `tsconfig.json`
1. Add `"jsx": "react"` to `tsconfig.json`
1. Run `./node_modules/.bin/tsc`

and after that I get the following error:
```
error TS2686: 'React' refers to a UMD global, but the current file is a module. Consider adding an import instead.
```

To fix it I use the following code in `index.tsx`:
```tsx
import * as React from "react";

interface ButtonProps extends React.HTMLAttributes<HTMLButtonElement> { }

export const Button: React.FunctionComponent<ButtonProps> = (props) => <button {...props} />;
```

(I just replaced named import with import the whole `React`).

After that your code compiles successfully.

But `dts-bundle-generator` works fine too: I run it with `./node_modules/.bin/dts-bundle-generator index.tsx --external-types react`. Here is output:

```ts
/// <reference types="react" />

export interface ButtonProps extends React.HTMLAttributes<HTMLButtonElement> {
}
export declare const Button: React.FunctionComponent<ButtonProps>;
```

Btw, see note for 2nd limitation [in the README](https://github.com/timocov/dts-bundle-generator#known-limitations) - it looks like this is exactly what you need 🙂

--- a/.gitignore
+++ b/.gitignore
@@ -7,6 +7,7 @@ dist/
 
 node_modules/
 !tests/test-cases/node_modules/
+!tests/test-cases/node_modules/**
 
 package-lock.json
 package-lock.json.*

--- a/src/bundle-generator.ts
+++ b/src/bundle-generator.ts
@@ -441,7 +441,12 @@ function isNodeUsed(
 	typeChecker: ts.TypeChecker
 ): boolean {
 	if (isNodeNamedDeclaration(node)) {
-		return rootFileExports.some((rootExport: ts.Symbol) => typesUsageEvaluator.isTypeUsedBySymbol(node, rootExport));
+		const nodeSymbol = getNodeSymbol(node, typeChecker);
+		if (nodeSymbol === null) {
+			return false;
+		}
+
+		return rootFileExports.some((rootExport: ts.Symbol) => typesUsageEvaluator.isSymbolUsedBySymbol(nodeSymbol, rootExport));
 	} else if (ts.isVariableStatement(node)) {
 		return node.declarationList.declarations.some((declaration: ts.VariableDeclaration) => {
 			return isNodeUsed(declaration, rootFileExports, typesUsageEvaluator, typeChecker);
@@ -484,4 +489,17 @@ function isSourceFileDefaultLibrary(program: ts.Program, file: ts.SourceFile): b
 
 	// tslint:disable-next-line:no-unnecessary-type-assertion
 	return (program as CompatibleProgram).isSourceFileDefaultLibrary(file);
+}
+
+function getNodeSymbol(node: ts.NamedDeclaration, typeChecker: ts.TypeChecker): ts.Symbol | null {
+	if (node.name === undefined) {
+		return null;
+	}
+
+	const symbol = typeChecker.getSymbolAtLocation(node.name);
+	if (symbol === undefined) {
+		return null;
+	}
+
+	return getActualSymbol(symbol, typeChecker);
 }

--- a/src/bundle-generator.ts
+++ b/src/bundle-generator.ts
@@ -113,21 +113,7 @@ export function generateDtsBundle(entries: ReadonlyArray<EntryPointConfig>, opti
 	const typeRoots = ts.getEffectiveTypeRoots(program.getCompilerOptions(), {});
 
 	const sourceFiles = program.getSourceFiles().filter((file: ts.SourceFile) => {
-		interface CompatibilityProgramPart {
-			// this method was introduced in TypeScript 2.6
-			// but to the public API it was added only in TypeScript 3.0
-			// so, to be compiled with TypeScript < 3.0 we need to have this hack
-			isSourceFileDefaultLibrary(file: ts.SourceFile): boolean;
-		}
-
-		type CommonKeys = keyof (CompatibilityProgramPart | ts.Program);
-
-		// if current ts.Program has isSourceFileDefaultLibrary method - then use it
-		// if it does not have it yet - use fallback
-		type CompatibleProgram = CommonKeys extends never ? ts.Program & CompatibilityProgramPart : ts.Program;
-
-		// tslint:disable-next-line:no-unnecessary-type-assertion
-		return !(program as CompatibleProgram).isSourceFileDefaultLibrary(file);
+		return !isSourceFileDefaultLibrary(program, file);
 	});
 
 	verboseLog(`Input source files:\n  ${sourceFiles.map((file: ts.SourceFile) => file.fileName).join('\n  ')}`);
@@ -480,4 +466,22 @@ function shouldNodeBeImported(node: ts.NamedDeclaration, rootFileExports: Readon
 
 		return rootFileExports.some((rootSymbol: ts.Symbol) => typesUsageEvaluator.isSymbolUsedBySymbol(symbol, rootSymbol));
 	});
+}
+
+function isSourceFileDefaultLibrary(program: ts.Program, file: ts.SourceFile): boolean {
+	interface CompatibilityProgramPart {
+		// this method was introduced in TypeScript 2.6
+		// but to the public API it was added only in TypeScript 3.0
+		// so, to be compiled with TypeScript < 3.0 we need to have this hack
+		isSourceFileDefaultLibrary(file: ts.SourceFile): boolean;
+	}
+
+	type CommonKeys = keyof (CompatibilityProgramPart | ts.Program);
+
+	// if current ts.Program has isSourceFileDefaultLibrary method - then use it
+	// if it does not have it yet - use fallback
+	type CompatibleProgram = CommonKeys extends never ? ts.Program & CompatibilityProgramPart : ts.Program;
+
+	// tslint:disable-next-line:no-unnecessary-type-assertion
+	return (program as CompatibleProgram).isSourceFileDefaultLibrary(file);
 }

--- a/src/bundle-generator.ts
+++ b/src/bundle-generator.ts
@@ -452,7 +452,7 @@ function isNodeUsed(
 }
 
 function shouldNodeBeImported(node: ts.NamedDeclaration, rootFileExports: ReadonlyArray<ts.Symbol>, typesUsageEvaluator: TypesUsageEvaluator): boolean {
-	const symbolsUsingNode = typesUsageEvaluator.getSymbolsUsingNode(node as ts.DeclarationStatement);
+	const symbolsUsingNode = typesUsageEvaluator.getSymbolsUsingNode(node);
 	if (symbolsUsingNode === null) {
 		throw new Error('Something went wrong - value cannot be null');
 	}

--- a/src/bundle-generator.ts
+++ b/src/bundle-generator.ts
@@ -156,7 +156,15 @@ export function generateDtsBundle(entries: ReadonlyArray<EntryPointConfig>, opti
 
 		const updateResultCommonParams = {
 			isStatementUsed: (statement: ts.Statement) => isNodeUsed(statement, rootFileExportSymbols, typesUsageEvaluator, typeChecker),
-			shouldStatementBeImported: (statement: ts.DeclarationStatement) => shouldNodeBeImported(statement, rootFileExportSymbols, typesUsageEvaluator),
+			shouldStatementBeImported: (statement: ts.DeclarationStatement) => {
+				return shouldNodeBeImported(
+					statement,
+					rootFileExportSymbols,
+					typesUsageEvaluator,
+					typeChecker,
+					isSourceFileDefaultLibrary.bind(null, program)
+				);
+			},
 			shouldDeclareGlobalBeInlined: (currentModule: ModuleInfo) => Boolean(outputOptions.inlineDeclareGlobals) && currentModule.type === ModuleType.ShouldBeInlined,
 			getModuleInfo: (fileName: string) => getModuleInfo(fileName, criteria),
 			getDeclarationsForExportedAssignment: (exportAssignment: ts.ExportAssignment) => {
@@ -456,8 +464,39 @@ function isNodeUsed(
 	return false;
 }
 
-function shouldNodeBeImported(node: ts.NamedDeclaration, rootFileExports: ReadonlyArray<ts.Symbol>, typesUsageEvaluator: TypesUsageEvaluator): boolean {
-	const symbolsUsingNode = typesUsageEvaluator.getSymbolsUsingNode(node);
+function shouldNodeBeImported(
+	node: ts.NamedDeclaration,
+	rootFileExports: ReadonlyArray<ts.Symbol>,
+	typesUsageEvaluator: TypesUsageEvaluator,
+	typeChecker: ts.TypeChecker,
+	isDefaultLibrary: (sourceFile: ts.SourceFile) => boolean
+): boolean {
+	const nodeSymbol = getNodeSymbol(node, typeChecker);
+	if (nodeSymbol === null) {
+		return false;
+	}
+
+	const symbolDeclarations = getDeclarationsForSymbol(nodeSymbol);
+	const isSymbolDeclaredInDefaultLibrary = symbolDeclarations.some(
+		(declaration: ts.Declaration) => isDefaultLibrary(declaration.getSourceFile())
+	);
+	if (isSymbolDeclaredInDefaultLibrary) {
+		// we shouldn't import a node declared in the default library (such dom, es2015)
+		// yeah, actually we should check that node is declared only in the default lib
+		// but it seems we can check that at least one declaration is from default lib
+		// to treat the node as un-importable
+		// because we can't re-export declared somewhere else node with declaration merging
+
+		// also, if some lib file will not be added to the project
+		// for example like it is described in the react declaration file (e.g. React Native)
+		// then here we still have a bug with "importing global declaration from a package"
+		// (see https://github.com/timocov/dts-bundle-generator/issues/71)
+		// but I don't think it is a big problem for now
+		// and it's possible that it will be fixed in https://github.com/timocov/dts-bundle-generator/issues/59
+		return false;
+	}
+
+	const symbolsUsingNode = typesUsageEvaluator.getSymbolsUsingSymbol(nodeSymbol);
 	if (symbolsUsingNode === null) {
 		throw new Error('Something went wrong - value cannot be null');
 	}

--- a/src/types-usage-evaluator.ts
+++ b/src/types-usage-evaluator.ts
@@ -19,12 +19,8 @@ export class TypesUsageEvaluator {
 		return this.isSymbolUsedBySymbolImpl(this.getActualSymbol(symbol), this.getActualSymbol(by), new Set<ts.Symbol>());
 	}
 
-	public getSymbolsUsingNode(typeNode: ts.NamedDeclaration): Set<ts.Symbol> | null {
-		if (typeNode.name === undefined) {
-			return null;
-		}
-
-		return this.nodesParentsMap.get(this.getSymbol(typeNode.name)) || null;
+	public getSymbolsUsingSymbol(symbol: ts.Symbol): Set<ts.Symbol> | null {
+		return this.nodesParentsMap.get(this.getActualSymbol(symbol)) || null;
 	}
 
 	private isSymbolUsedBySymbolImpl(fromSymbol: ts.Symbol, toSymbol: ts.Symbol, visitedSymbols: Set<ts.Symbol>): boolean {

--- a/src/types-usage-evaluator.ts
+++ b/src/types-usage-evaluator.ts
@@ -15,15 +15,6 @@ export class TypesUsageEvaluator {
 		this.computeUsages(files);
 	}
 
-	public isTypeUsedBySymbol(typeNode: ts.NamedDeclaration, by: ts.Symbol): boolean {
-		if (typeNode.name === undefined) {
-			// anon?
-			return false;
-		}
-
-		return this.isSymbolUsedBySymbol(this.getSymbol(typeNode.name), by);
-	}
-
 	public isSymbolUsedBySymbol(symbol: ts.Symbol, by: ts.Symbol): boolean {
 		return this.isSymbolUsedBySymbolImpl(this.getActualSymbol(symbol), this.getActualSymbol(by), new Set<ts.Symbol>());
 	}

--- a/tests/functional-test-cases.ts
+++ b/tests/functional-test-cases.ts
@@ -31,10 +31,14 @@ function getTestCases(): TestCase[] {
 		.map((directoryName: string) => {
 			const testCaseDir = path.resolve(testCasesDir, directoryName);
 			const outputFileName = path.resolve(testCaseDir, 'output.d.ts');
+			const inputFileName = path.relative(process.cwd(), path.resolve(testCaseDir, 'input.ts'));
+
+			assert(fs.existsSync(inputFileName), `Input file doesn't exist for ${directoryName}`);
+			assert(fs.existsSync(outputFileName), `Output file doesn't exist for ${directoryName}`);
 
 			const result: TestCase = {
 				name: directoryName,
-				inputFileName: path.relative(process.cwd(), path.resolve(testCaseDir, 'input.ts')),
+				inputFileName,
 				config: require(path.resolve(testCaseDir, 'config.js')) as TestCaseConfig,
 				outputFileContent: prepareString(fs.readFileSync(outputFileName, 'utf-8')),
 			};

--- a/tests/test-cases/import-package-with-declaration-merging-with-default-lib/config.ts
+++ b/tests/test-cases/import-package-with-declaration-merging-with-default-lib/config.ts
@@ -1,0 +1,9 @@
+import { TestCaseConfig } from '../test-case-config';
+
+const config: TestCaseConfig = {
+	libraries: {
+		importedLibraries: ['fake-types-lib-3'],
+	},
+};
+
+export = config;

--- a/tests/test-cases/import-package-with-declaration-merging-with-default-lib/input.ts
+++ b/tests/test-cases/import-package-with-declaration-merging-with-default-lib/input.ts
@@ -1,0 +1,3 @@
+import { FooBar } from 'fake-types-lib-3';
+
+export type Baz = FooBar<ArrayConstructor>;

--- a/tests/test-cases/import-package-with-declaration-merging-with-default-lib/output.d.ts
+++ b/tests/test-cases/import-package-with-declaration-merging-with-default-lib/output.d.ts
@@ -1,0 +1,3 @@
+import { FooBar } from 'fake-types-lib-3';
+
+export declare type Baz = FooBar<ArrayConstructor>;

